ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+	}()
+
+	writer := NewCRC32Writer(closable)
+
+	topLevelNodeIndices, topLevelNodesCount, err := storeTopLevelNodes(
+		tries,
+		subTrieRootIndices,
+		subTriesNodeCount+1, // the counter is 1 more than the node count, because the first item is nil
+		writer)
+
+	if err != nil {
+		return 0, fmt.Errorf("could not store top level nodes: %w", err)
+	}
+
+	logger.Info().Msgf("top level nodes have been stored. top level node count: %v", topLevelNodesCount)
+
+	err = storeRootNodes(tries, topLevelNodeIndices, flattener.EncodeTrie, writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store top level nodes: %w", err)
+	}
+
+	err = storeFooter(topLevelNodesCount, uint16(len(tries)), writer)
+	if err != nil {
+		return 0, fmt.Errorf("could not store footer: %w", err)
+	}
+
+	// write checksum to the end of the file
+	checksum := writer.Crc32()
+	_, err = writer.Write(encodeCRC32Sum(checksum))
+	if err != nil {
+		return 0, fmt.Errorf("cannot write CRC32 checksum to top level part file: %w", err)
+	}
+	return checksum, nil
+}
+
+func createSubTrieRoots(tries []*trie.MTrie) [subtrieCount][]*node.Node {
+	var subtrieRoots [subtrieCount][]*node.Node
+	for i := 0; i < len(subtrieRoots); i++ {
+		subtrieRoots[i] = make([]*node.Node, len(tries))
+	}
+
+	for trieIndex, t := range tries {
+		// subtries is an array with subtrieCount trie nodes
+		// in breadth-first order at subtrieLevel of the trie `t`
+		subtries := getNodesAtLevel(t.RootNode(), subtrieLevel)
+		for subtrieIndex, subtrieRoot := range subtries {
+			subtrieRoots[subtrieIndex][trieIndex] = subtrieRoot
+		}
+	}
+	return subtrieRoots
+}
+
+func estimateSubtrieNodeCount(tries []*trie.MTrie) int {
+	if len(tries) == 0 {
+		return 0
+	}
+	estimatedTrieNodeCount := 2*int(tries[0].AllocatedRegCount()) - 1
+	return estimatedTrieNodeCount / subtrieCount
+}
+
+func storeSubTrieConcurrently(
+	subtrieRoots [subtrieCount][]*node.Node,
+	estimatedSubtrieNodeCount int,
+	outputDir string,
+	outputFile string,
+	logger *zerolog.Logger,
+) (
+	map[*node.Node]uint64, // node indices
+	uint64, // node count
+	[]uint32, //checksums
+	error, // any exception
+) {
+	logger.Info().Msgf("storing %v subtrie groups with average node count %v for each subtrie", subtrieCount, estimatedSubtrieNodeCount)
+
+	resultChs := make([]chan *resultStoringSubTrie, 0, len(subtrieRoots))
+	for i, subTrieRoot := range subtrieRoots {
+		resultCh := make(chan *resultStoringSubTrie)
+		go func(i int, subTrieRoot []*node.Node) {
+			roots, nodeCount, checksum, err := storeCheckpointSubTrie(
+				i, subTrieRoot, estimatedSubtrieNodeCount, outputDir, outputFile, logger)
+			resultCh <- &resultStoringSubTrie{
+				Index:     i,
+				Roots:     roots,
+				NodeCount: nodeCount,
+				Checksum:  checksum,
+				Err:       err,
+			}
+			close(resultCh)
+		}(i, subTrieRoot)
+		resultChs = append(resultChs, resultCh)
+	}
+
+	logger.Info().Msgf("subtrie roots have been stored")
+
+	results := make(map[*node.Node]uint64, 1<<(subtrieLevel+1))
```

Comment:
Did you mean `subtrieCount+1`?

Reply:
I didn't change this part, it's [the same as before ](https://github.com/onflow/flow-go/blob/master/ledger/complete/wal/checkpointer.go#L355). Maybe might know why it's this value.

This also doesn't matter too much, we are allocating memory for the root nodes at subtrie level, which usually only have at most hundreds of nodes.

Reply:
@fxamacker why isn't `subtrieCount * len(subtrieRoots[0])`?

If there is only one trie, since there are at most subtrieCount number of nodes at level `subtrieLevel`, then for `subtrieRoots[0]` number of tries, there will be `subtrieCount * subtrieRoots[0]` number of nodes at most.

The number of top level nodes is nearly the same as total number of nodes at subtrie level. right?

Reply:
@zhangchiqing Although what you say is right about this specific block of code, returned `results` map is used by other function which adds more elements.

In this function `storeSubTrieConcurrently()`, `results` only contains subtrie root nodes for all tries, which is `subtrieCount * subtrieRoots[0]` number of nodes.

However, its caller `StoreCheckpointV6()` [passes returned `results` map to `storeTopLevelNodesAndTrieRoots()` as function parameter](https://github.com/onflow/flow-go/blob/db755dea37553d61ba1ca6432e1cc2c3142b5f51/ledger/complete/wal/checkpoint_v6_writer.go#L58-L75).

`storeTopLevelNodesAndTrieRoots()` continues to add top level nodes (from level 0 to including level 3) to this map.  So this map contains top level nodes (from 0 to including level 4) for all tries at the end.

Reply:
Good point